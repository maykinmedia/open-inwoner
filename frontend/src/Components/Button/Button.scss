.button {
  align-items: center;
  border-radius: 3px;
  box-sizing: border-box;
  color: var(--color-secondary);
  cursor: pointer;
  display: inline-flex;
  font-family: var(--font-family-body);
  font-size: var(--font-size-body);
  letter-spacing: 0;
  line-height: var(--font-line-height-body);
  padding: var(--spacing-medium) var(--spacing-large);
  text-align: center;
  text-decoration: none;
  transition: transform 300ms, background-color 300ms;

  &:focus,
  &:hover {
    filter: brightness(80%);
    transform: translateY(-1px);
  }

  &--icon-before {
    *[class*=icon],
    *[class*=Icon] {
      margin-right: var(--spacing-medium)
    }
  }

  &--icon-after {
    flex-direction: row-reverse;
    //padding-right: 0;

    *[class*=icon],
    *[class*=Icon] {
      margin-left: var(--spacing-medium)
    }
  }

  &--big {
    height: var(--row-height-big);

    @media (max-width: 767px) {
      width: 100%;
      justify-content: space-between;
    }
  }

  &--open {
    background-color: transparent;
    color: var(--color-primary);
    border-color: transparent;

    &:focus,
    &:hover {
      border-color: var(--color-primary);
    }
  }

  &--transparent {
    background-color: transparent;
    border-color: transparent;
    padding-left: 0;
    padding-right: 0;
  }

  &--transparent#{&}--icon-before {
  }

  &--transparent#{&}--icon-after {
  }

  &--primary {
    background-color: var(--color-primary);
    border: 1px solid var(--color-primary);
    color: var(--color-white);
  }
}

.h4 + .button,
.p + .button {
  margin-top: var(--spacing-large)
}
