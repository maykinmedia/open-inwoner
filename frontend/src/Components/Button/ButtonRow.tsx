import React, { ReactElement} from 'react';

import './ButtonRow.scss'

interface iButtonRowProps {
  startingText: string,
  buttons: Array<ReactElement>,
}

export function ButtonRow(props: iButtonRowProps) {
  return (
    <div className="button-row">
      { props.startingText }
      { props.buttons }
    </div>
  );
}
