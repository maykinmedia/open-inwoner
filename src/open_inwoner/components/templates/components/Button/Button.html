{% load helpers icon_tags %}

{% if href %}
    <a class="{{ classes }}" href="{{href}}" {% if href|startswith:"http" %}target="_blank"{% endif %}{% if hide_text %} aria-describedby="{{ text }}" title="{{ text }}"{% endif %}>
        {% if icon %}{% icon icon=icon outlined=icon_outlined %}{% endif %}
        {% if not hide_text %}{{text}}{% endif %}
    </a>
{% else %}
    <button class="{{ classes }}" type="{{type}}" name="{{ name }}">
        {% if icon %}{% icon icon=icon outlined=icon_outlined %}{% endif %}
        {% if not hide_text %}{{text}}{% endif %}
    </button>
{% endif %}
