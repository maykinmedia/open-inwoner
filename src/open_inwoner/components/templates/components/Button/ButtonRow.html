<div class="button-row {% if align == "right" %}button-row--right{% endif %}">
    {{ contents }}
</div>
