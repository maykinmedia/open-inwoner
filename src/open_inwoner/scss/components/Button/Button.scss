.button {
  align-items: center;
  border-radius: var(--border-radius);
  box-sizing: border-box;
  color: var(--color-secondary);
  background-color: transparent;
  cursor: pointer;
  display: inline-flex;
  gap: var(--spacing-medium);
  font-family: var(--font-family-body);
  font-size: var(--font-size-body);
  letter-spacing: 0;
  height: var(--row-height);
  line-height: var(--font-line-height-body);
  padding: var(--spacing-medium) var(--spacing-large);
  text-align: center;
  text-decoration: none;
  transition: transform 300ms, background-color 300ms;
  font-weight: normal;
  // modified outline for accessibility focus
  outline-width: 1px;
  outline-offset: 1px;
  border: 0;
  white-space: nowrap;

  &:focus,
  &:hover {
    transform: translateY(-1px);
    box-shadow: none;
  }

  *[class*='icon'],
  *[class*='Icon'] {
    font-size: 20px;
    max-width: 20px;
  }

  &--icon-after {
    flex-direction: row-reverse;
  }

  &--textless {
    *[class*='icon'],
    *[class*='Icon'] {
      margin: 0;
    }
  }

  &--small {
    font-size: var(--font-size-body-small);
    line-height: var(--font-line-height-body-small);
    padding: var(--spacing-medium) var(--spacing-medium);
    gap: var(--spacing-small);
    height: var(--row-height-small);
  }

  &--big {
    height: var(--row-height-big);

    @media (max-width: 767px) {
      width: 100%;
      justify-content: space-between;
    }
  }

  &--fullwidth {
    display: block;
    width: 100%;
  }

  &--spaceless {
    padding: 0;
  }

  &#{&}--primary {
    background-color: var(--color-primary);
    border: 1px solid var(--color-primary);
    color: var(--color-font-primary);

    &:focus,
    &:hover {
      background-color: var(--color-primary-darker);
    }
  }

  &--secondary {
    background-color: var(--color-secondary);
    border: 1px solid var(--color-secondary);
    color: var(--color-white);

    &:focus,
    &:hover {
      background-color: var(--color-secondary-darker);
      color: var(--color-font-primary);
    }
  }

  &--transparent {
    background-color: transparent;
    border-color: transparent;
    color: var(--color-primary);

    &.button--primary {
      border-color: transparent;
      color: var(--color-primary);
      background-color: transparent;

      &:hover {
        background-color: transparent;
        border-color: transparent;
        color: var(--color-primary);
        text-decoration: underline;
      }
    }

    &.button--secondary {
      color: var(--color-secondary);
      background-color: transparent;
    }
  }

  &--open {
    background-color: transparent;
    color: var(--color-primary);
    border: 1px solid transparent;

    &:focus,
    &:hover {
      border-color: var(--color-primary);
    }
  }

  &--bordered {
    background-color: transparent;
    border: 1px solid var(--color-primary);
    color: var(--color-primary);
  }

  &--pill {
    background-color: #fff;
    position: relative;
    color: var(--color-primary);
    filter: saturate(50%);
    border-radius: 100px; // This is big to make sure the corners are rounded properly.

    &:before {
      content: ' ';
      border-radius: 100px; // This is big to make sure the corners are rounded properly.
      position: absolute;
      top: 0;
      bottom: 0;
      left: 0;
      right: 0;
      background-color: var(--color-primary);
      opacity: 0.15;
    }
  }

  &--disabled {
    border-color: var(--color-gray) !important;
    color: var(--color-gray-light);
    pointer-events: none;
    cursor: default;
  }

  &--danger {
    background-color: var(--color-danger-lighter);
    color: var(--color-white);

    &:hover {
      transform: translateY(-1px);
      background-color: var(--color-danger);
    }
  }

  &--danger-borderless {
    color: var(--color-danger-lighter);
  }

  &--error {
    color: var(--color-error-darker);
  }

  &--primary-close {
    background-color: var(--color-primary);
    border: 1px solid var(--color-primary);
    color: var(--color-font-primary);
  }

  &--error-confirm {
    background-color: var(--color-error-darker);
    color: var(--color-white);
  }

  &--borderless {
    background-color: rgba(0, 0, 0, 0);
    border: none;
    color: var(--color-gray-dark);
    padding: 0;
  }

  &__text-wrapper {
    display: flex;
  }

  > .link__text {
    width: 100%;
  }

  &[hidden] {
    display: none;
  }
}

.utrecht-heading-2 + .button,
.utrecht-heading-4 + .button,
.utrecht-paragraph + .button {
  margin-top: var(--spacing-large);
}

.cta-button {
  margin-bottom: var(--spacing-large);
}

///
/// Legacy styling
///

.h2 + .button,
.h4 + .button,
.p + .button {
  margin-top: var(--spacing-large);
}
