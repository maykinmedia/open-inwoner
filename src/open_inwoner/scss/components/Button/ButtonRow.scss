.button-row {
  display: flex;
  align-items: center;
  gap: var(--spacing-medium);
  margin: var(--spacing-medium) 0;
  flex-direction: row;

  &--right {
    flex-direction: row-reverse;
    justify-content: end;
  }

  &--mobile-only {
    @media (min-width: 768px) {
      display: none;
    }
  }
}
