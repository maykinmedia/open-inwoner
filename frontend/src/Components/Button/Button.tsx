import React, {ComponentType} from 'react';
import {Link} from 'react-router-dom';
import './Button.scss';


export interface iButtonProps {
  href?: string,
  icon?: ComponentType,
  iconPosition?: 'before' | 'after' | undefined,
  size?: 'big' | 'normal' | undefined,
  type?: 'button' | 'submit' | 'reset' | undefined,
  open?: boolean,
  primary?: boolean
  secondary?: boolean
  transparent?: boolean,
  children?: any,
  [key: string]: any,
}


/**
 * A generic button, can be a link or a button.
 * @param {iButtonProps} props
 * @return {ReactElement}
 */
export function Button(props: iButtonProps) {
  const {href, icon, iconPosition, size, type, open, primary, secondary, transparent, children, ..._props} = props;

  /**
   * Returns the className value.
   * @return {string}
   */
  const getClassNames = () => {
    let classNames = 'button';

    if (icon) {
      classNames += ' button--icon'
    }

    if (iconPosition) {
      classNames += ` button--icon-${iconPosition}`
    }

    if (size) {
      classNames += ` button--${size}`
    }

    if (open) {
      classNames += ' button--open';
    }

    if (primary) {
      classNames += ' button--primary';
    }

    if (secondary) {
      classNames += ' button--secondary';
    }

    if (transparent) {
      classNames += ' button--transparent';
    }
    return classNames;
  };

  /**
   * Returns the children.
   * @return {ReactElement}
   */
  const getChildren = () => {
    const Icon = icon;

    return (
      <>
        {Icon && <Icon/>}
        {children}
      </>
    );
  };

  if (href) {
    if (href.startsWith('http')) {
      return <a className={getClassNames()} href={href} {..._props}>{getChildren()}</a>;
    }
    return <Link className={getClassNames()} to={href} {..._props}>{getChildren()}</Link>;
  }
  return <button className={getClassNames()} type={type} {..._props}>{getChildren()}</button>;
}

Button.defaultProps = {
  iconPosition: 'before',
}
