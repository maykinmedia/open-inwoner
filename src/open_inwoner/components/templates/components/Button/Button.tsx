import React, {ComponentType} from 'react';
import { Link } from 'react-router-dom';
import CloseIcon from '@mui/icons-material/Close';
import './Button.scss';


export interface iButtonProps {
  href?: string,
  icon?: ComponentType,
  iconPosition?: 'before' | 'after' | undefined,
  size?: 'big' | 'normal' | undefined,
  type?: 'button' | 'submit' | 'reset' | undefined,
  open?: boolean,
  bordered?: boolean,
  primary?: boolean
  secondary?: boolean
  transparent?: boolean,
  children?: any,
  closeAction?: Function,
  [key: string]: any,
}


/**
 * A generic button, can be a link or a button.
 * @param {iButtonProps} props
 * @return {ReactElement}
 */
export function Button(props: iButtonProps) {
  const {href, icon, iconPosition, size, type, open, bordered, primary, secondary, transparent, children, closeAction, ..._props} = props;

  /**
   * Returns the className value.
   * @return {string}
   */
  const getClassNames = () => {
    let classNames = 'button';

    if (icon) {
      classNames += ' button--icon'
    }

    if (iconPosition) {
      classNames += ` button--icon-${iconPosition}`
    }

    if (size) {
      classNames += ` button--${size}`
    }

    if (open) {
      classNames += ' button--open';
    }

    if (bordered) {
      classNames += ' button--bordered';
    }

    if (primary) {
      classNames += ' button--primary';
    }

    if (secondary) {
      classNames += ' button--secondary';
    }

    if (transparent) {
      classNames += ' button--transparent';
    }
    return classNames;
  };

  /**
   * Returns the children.
   * @return {ReactElement}
   */
  const getChildren = () => {
    const Icon = icon;

    return (
      <>
        {Icon && <Icon/>}
        {closeAction && <CloseIcon /> }
        {children}
      </>
    );
  };

  if (href) {
    if (href.startsWith('http')) {
      return <a className={getClassNames()} href={href} {..._props}>{getChildren()}</a>;
    }
    return <Link className={getClassNames()} to={href} {..._props}>{getChildren()}</Link>;
  }
  return <button onClick={ closeAction } className={getClassNames()} type={type} {..._props}>{getChildren()}</button>;
}

Button.defaultProps = {
  iconPosition: 'before',
}
