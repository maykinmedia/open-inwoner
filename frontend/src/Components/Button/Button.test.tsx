import React from 'react';
import renderer from 'react-test-renderer';
import { Button } from './Button';
import ArrowForwardIcon from '@mui/icons-material/ArrowForward';
import { Router } from 'react-router-dom';

test('Render button', () => {
  const component = renderer.create(
    <Button>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

// TODO: Render this button
// test('Render link button', () => {
//   const component = renderer.create(
//     <Router>
//       <Button href="/maykin">Maykin</Button>,
//     </Router>
//   );
//   let tree = component.toJSON();
//   expect(tree).toMatchSnapshot();
// });

test('Render anchor button', () => {
  const component = renderer.create(
    <Button href="http://www.maykin.nl">Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with icon', () => {
  const component = renderer.create(
    <Button icon={ArrowForwardIcon}>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with size', () => {
  const component = renderer.create(
    <Button size="big">Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with open', () => {
  const component = renderer.create(
    <Button open={ true }>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with primary', () => {
  const component = renderer.create(
    <Button primary={ true }>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with secondary', () => {
  const component = renderer.create(
    <Button secondary={ true }>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with transparent', () => {
  const component = renderer.create(
    <Button transparent={ true }>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button without iconPosition', () => {
  const component = renderer.create(
    <Button iconPosition={ undefined }>Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});

test('Render button with iconPosition', () => {
  const component = renderer.create(
    <Button iconPosition="after">Maykin</Button>,
  );
  let tree = component.toJSON();
  expect(tree).toMatchSnapshot();
});
