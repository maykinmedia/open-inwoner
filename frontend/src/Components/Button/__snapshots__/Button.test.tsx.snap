// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Render anchor button 1`] = `
<a
  className="button button--icon-after"
  href="http://www.maykin.nl"
>
  Maykin
</a>
`;

exports[`Render button 1`] = `
<button
  className="button button--icon-after"
>
  Maykin
</button>
`;

exports[`Render button with icon 1`] = `
<button
  className="button button--icon button--icon-after"
>
  <svg
    aria-hidden={true}
    className="MuiSvgIcon-root MuiSvgIcon-fontSizeMedium css-i4bv87-MuiSvgIcon-root"
    data-testid="ArrowForwardIcon"
    focusable="false"
    viewBox="0 0 24 24"
  >
    <path
      d="m12 4-1.41 1.41L16.17 11H4v2h12.17l-5.58 5.59L12 20l8-8z"
    />
  </svg>
  Maykin
</button>
`;

exports[`Render button with iconPosition 1`] = `
<button
  className="button button--icon-after"
>
  Maykin
</button>
`;

exports[`Render button with open 1`] = `
<button
  className="button button--icon-after button--open"
>
  Maykin
</button>
`;

exports[`Render button with primary 1`] = `
<button
  className="button button--icon-after button--primary"
>
  Maykin
</button>
`;

exports[`Render button with secondary 1`] = `
<button
  className="button button--icon-after button--secondary"
>
  Maykin
</button>
`;

exports[`Render button with size 1`] = `
<button
  className="button button--icon-after button--big"
>
  Maykin
</button>
`;

exports[`Render button with transparent 1`] = `
<button
  className="button button--icon-after button--transparent"
>
  Maykin
</button>
`;

exports[`Render button without iconPosition 1`] = `
<button
  className="button button--icon-after"
>
  Maykin
</button>
`;
